---
title: "PLS_of_proteome_transcriptome_metabolome"
author: "Devin Molnau"
date: "May 7, 2019"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(doParallel)
library(dplyr)
library(ggplot2)
library(grid)
library(igraph)
library(NetRep)
library(reshape2)
library(scales)
library(tidyverse)
library(GENIE3)
library(mixOmics)
source("./src/bcb570_utils.R")
theme_set(theme_light())
```
# PCA Analysis

### Read in Data Sets

Reads in the proteomics, metabolomics, and transcriptomics data as dataframes.

```{r}
proteomics <- read.delim("./data/raw/Proteomics2.txt")
metabolomics<-read.delim("./data/raw/Metabolomics2.txt")
transcriptomics<-read.delim("./data/raw/Transcriptomics2.txt")
```

### PCA of just proteome

```{r}
prot_genotypes<-t(proteomics[8:16])
colnames(prot_genotypes) <- proteomics$UNIQID
pca.proteome <- pca(prot_genotypes, ncomp = 6, center = TRUE, scale = TRUE)
pca.proteome
plot(pca.proteome)
title(main = "PCA Plot: Proteomics")
```

```{r}
groups<-data.frame(genotypes=c("Histidine mutant", "Histidine mutant","Histidine mutant","Leucine mutant","Leucine mutant","Leucine mutant","WT","WT","WT"),
                   replicates=c("H1","H2","H3","L1","L2","L3","PROTOTROPH1","PROTOTROPH2","PROTOTROPH3")
                   )

plotIndiv(pca.proteome, comp = c(1, 2), group = groups$genotypes,
          ind.names = groups$replicates,
          legend = TRUE, title = 'Proteome, PCA comp 1 - 2') 

```

What genes are in the first principal component?

```{r}
pca_nth_component_histogram(pca.proteome, 1) 
```

```{r}
names(get_top_loadings(pca.proteome, 1))
```


### PCA of just transcriptome

When you attempt to run the pca on the transcriptome alone, there is an error due to none of the genotypes expressing that gene (aka that column is filled with just zeros).

```{r, error = TRUE}
trans_genotypes<-t(transcriptomics[6:14])
colnames(trans_genotypes)<-transcriptomics$Unique_ID
trans_genotypes <- trans_genotypes[, colSums(trans_genotypes != 0) > 0]
pca.transcriptome <- pca(trans_genotypes, ncomp = 6, center = TRUE, scale = TRUE) 
pca.transcriptome

plot(pca.transcriptome) + title(main="PCA Plot: Transcriptomics")
```



```{r}
names(get_top_loadings(pca.transcriptome, 1)) %>% 
  write.table(file = "transcriptome_comp1", row.names = FALSE, col.names = FALSE,
              sep = "\n", quote = FALSE)
```


### Merging data

Scaling numeric data in both the proteomic and transcriptomic datasets:


```{r}
proteomic.transcriptomic.merged <- scale_and_merge_prot_transc(proteomics, transcriptomics)
pca.proteome_transcriptome <- pca(proteomic.transcriptomic.merged, ncomp = 10, center = TRUE, scale = TRUE)

pca.proteome_transcriptome

plot(pca.proteome_transcriptome) + title(main="PCA Plot: Proteomic and transcriptomic data")

```

Let's see what is in the first principal component.


```{r}
pca_nth_component_histogram(pca.proteome_transcriptome, 1)
```

```{r}
get_top_loadings(pca.proteome_transcriptome, 2) %>% names %>% 
  write.table(file = "proteome_transcriptome.pca.comp_2", row.names = FALSE, col.names = FALSE,
              sep = "\n", quote = FALSE)
```



```{r}
groups<-data.frame(genotypes=c("Histidine mutant", "Histidine mutant","Histidine mutant","Leucine mutant","Leucine mutant","Leucine mutant","WT","WT","WT","WT","WT","WT","Histidine mutant", "Histidine mutant","Histidine mutant","Leucine mutant","Leucine mutant","Leucine mutant") ,replicates=c("H1","H2","H3","L1","L2","L3","PROTOTROPH1","PROTOTROPH2","PROTOTROPH3"))

plotIndiv(pca.proteome_transcriptome, comp = c(1, 2), group = groups$genotypes,
          ind.names = rownames(prot_tran_genotypes),
          legend = TRUE, title = 'Proteome + Transcriptome PCA, comps 1 & 2')
```

```{r}
plotIndiv(pca.proteome_transcriptome, comp = c(1,2), group = groups$genotypes,
          ind.names = rownames(prot_tran_genotypes),
          legend = TRUE, title = 'Proteome and Transcriptome, PCA comp 2 - 3')
```


### PCA of proteomic, transcripmtomic, and metabolomic data
```{r}
prot_tran_met_genotypes<-t(cbind(prot_met_trans[9:17],prot_met_trans[21:29],prot_met_trans[31:41]))
pca.proteome_transcriptome_metabolome <- pca(prot_tran_met_genotypes, ncomp = 10, center = TRUE, scale = TRUE)

pca.proteome_transcriptome_metabolome

plot(pca.proteome_transcriptome_metabolome)
title(main="Proteomic,Transcriptomic,Metabolomic PCA Plot")
```

```{r}
groups<-data.frame(genotypes=c("Histidine mutant", "Histidine mutant","Histidine mutant","Leucine mutant","Leucine mutant","Leucine mutant","WT","WT","WT","WT","WT","WT","Histidine mutant", "Histidine mutant","Histidine mutant","Leucine mutant","Leucine mutant","Leucine mutant","Histidine mutant","Histidine mutant", "Histidine mutant","Histidine mutant","Leucine mutant","Leucine mutant","Leucine mutant","WT","WT","WT","WT"))
length(groups$genotypes)
rownames(prot_tran_met_genotypes)
plotIndiv(pca.proteome_transcriptome_metabolome, comp = c(1,2), group = groups$genotypes,
          ind.names = rownames(prot_tran_met_genotypes),
          legend = TRUE, title = 'Proteome,Transcriptome, Metabolome, PCA comp 1 - 2')
```

# sPLS Analysis
```{r}
prot_trans.spls <- spls( prot_genotypes, trans_genotypes, ncomp = 6, keepX = c(20,20,20), keepY= c(10,10,10), mode = "regression")
tune.spls <- perf(prot_trans.spls, validation = "Mfold", folds = 6, progressBar = FALSE, nrepeat = 50)




plot(tune.spls$Q2.total) + abline(h = 0.0975)

tune.spls$Q2.total


```
### Sample plots
```{r}
par(mfrow = c(1,3))
groups<-data.frame(genotypes=c("Histidine mutant", "Histidine mutant","Histidine mutant","Leucine mutant","Leucine mutant","Leucine mutant","WT","WT","WT") ,replicates=c("H1","H2","H3","L1","L2","L3","PROTOTROPH1","PROTOTROPH2","PROTOTROPH3"))

plotIndiv(prot_trans.spls, comp = 5:6, rep.space= 'Y-variate', group = groups$genotypes,
          ind.names = genotype$replicates,
          legend = TRUE, title = 'Proteome and Transcriptome, sPLS comp 1 - 2, Y-space')


par(mfrow = c(1,3))

```

```{r}
plotIndiv(prot_trans.spls, comp = 1:2, rep.space= 'X-variate', group = groups$genotypes,
          ind.names = groups$replicates,
          legend = TRUE, title = 'Liver, sPLS comp 1 - 2, X-space')

plotIndiv(prot_trans.spls, comp = 1:2, rep.space= 'XY-variate', group = groups$genotypes,
          ind.names = groups$replicates,
          legend = TRUE, title = 'Liver, sPLS comp 1 - 2, XY-space')
```

```{r}

plotVar(prot_trans.spls, comp =1:2, 
        var.names = list(X.label = proteomics$UNIQID, 
                         Y.label = TRUE), cex = c(4, 5))
  color.edge <- color.GreenRed(50) 

```

