---
title: "PLS_of_proteome_transcriptome_metabolome"
author: "Devin Molnau"
date: "May 7, 2019"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(doParallel)
library(dplyr)
library(ggplot2)
library(grid)
library(igraph)
library(NetRep)
library(reshape2)
library(scales)
library(tidyverse)
library(GENIE3)
library(mixOmics)
setwd("D:/Documents/BoxSync/Classes/BCB570/final\ project/my_work/")
theme_set(theme_light())
```

## Read in Data Sets

Reads in the proteomics, metabolomics, and transcriptomics data as dataframes.

```{r}
proteomics <- read.delim("./data/raw/Proteomics2.txt")
metabolomics<-read.delim("./data/raw/Metabolomics2.txt")
transcriptomics<-read.delim("./data/raw/Transcriptomics2.txt")
```

## PCA of just proteome
```{r}
prot_genotypes<-t(proteomics[8:16])
colnames(prot_genotypes)<-proteomics$UNIQID
pca.proteome <- pca(prot_genotypes, ncomp = 6, center = TRUE, scale = TRUE)
pca.proteome
plot(pca.proteome)
title(main="Protein PCA Plot")
```

```{r}
groups<-data.frame(genotypes=c("Histidine mutant", "Histidine mutant","Histidine mutant","Leucine mutant","Leucine mutant","Leucine mutant","WT","WT","WT"),replicates=c("H1","H2","H3","L1","L2","L3","PROTOTROPH1","PROTOTROPH2","PROTOTROPH3"))

plotIndiv(pca.proteome, comp = c(1,2), group = groups$genotypes,
          ind.names = groups$replicates,
          legend = TRUE, title = 'Proteome, PCA comp 1 - 2')

```

## PCA of just transcriptome
When you attempt to run the pca on the transcriptome alone, there is an error due to none of the genotypes expressing that gene (aka that column is filled with just zeros).
```{r}
trans_genotypes<-t(transcriptomics[6:14])
colnames(trans_genotypes)<-transcriptomics$Unique_ID
pca.transcriptome <- pca(trans_genotypes, ncomp = 6, center = TRUE, scale = TRUE) #Error: cannot rescale a constant/zero column to unit variance.
pca.transcriptome

plot(pca.transcriptome)
title(main="transcriptomics PCA Plot")
```


Attempt to merge transcriptome and proteome data
```{r}
proteome_transcriptome<-merge(proteomics, transcriptomics, by.y="Unique_ID", by.x = "UNIQID")
prot_met_trans<-merge(proteome_transcriptome,metabolomics,by.x = "Pathway", by.y="Pathway", all.x = FALSE)
```

#Attempt at PCA of combined proteomics and transcriptomics data
```{r}
proteome_transcriptome[8:16]
prot_tran_genotypes<-t(cbind(proteome_transcriptome[8:16],proteome_transcriptome[21:29]))
colnames(prot_tran_genotypes)<-proteome_transcriptome$UNIQID
pca.proteome_transcriptome <- pca(prot_tran_genotypes, ncomp = 10, center = TRUE, scale = TRUE)

pca.proteome_transcriptome

plot(pca.proteome_transcriptome)
title(main="Protein and Transcriptomic PCA Plot")

```

```{r}
groups<-data.frame(genotypes=c("Histidine mutant", "Histidine mutant","Histidine mutant","Leucine mutant","Leucine mutant","Leucine mutant","WT","WT","WT","WT","WT","WT","Histidine mutant", "Histidine mutant","Histidine mutant","Leucine mutant","Leucine mutant","Leucine mutant") ,replicates=c("H1","H2","H3","L1","L2","L3","PROTOTROPH1","PROTOTROPH2","PROTOTROPH3"))

plotIndiv(pca.proteome_transcriptome, comp = c(1,2), group = groups$genotypes,
          ind.names = rownames(prot_tran_genotypes),
          legend = TRUE, title = 'Proteome and Transcriptome, PCA comp 1 - 2')
```

## PCA of proteomic, transcripmtomic, and metabolomic data
```{r}
prot_tran_met_genotypes<-t(cbind(prot_met_trans[9:17],prot_met_trans[21:29],prot_met_trans[31:41]))
pca.proteome_transcriptome_metabolome <- pca(prot_tran_met_genotypes, ncomp = 10, center = TRUE, scale = TRUE)

pca.proteome_transcriptome_metabolome

plot(pca.proteome_transcriptome_metabolome)
title(main="Protein, Transcriptomic,Metabolomic PCA Plot")
```

```{r}
groups<-data.frame(genotypes=c("Histidine mutant", "Histidine mutant","Histidine mutant","Leucine mutant","Leucine mutant","Leucine mutant","WT","WT","WT","WT","WT","WT","Histidine mutant", "Histidine mutant","Histidine mutant","Leucine mutant","Leucine mutant","Leucine mutant","Histidine mutant","Histidine mutant", "Histidine mutant","Histidine mutant","Leucine mutant","Leucine mutant","Leucine mutant","WT","WT","WT","WT"))
length(groups$genotypes)
rownames(prot_tran_met_genotypes)
plotIndiv(pca.proteome_transcriptome_metabolome, comp = c(1,2), group = groups$genotypes,
          ind.names = rownames(prot_tran_met_genotypes),
          legend = TRUE, title = 'Proteome and Transcriptome, PCA comp 1 - 2')
```

