---
title: "EDA"
author: "Paul Villanueva"
date: "5/1/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(dplyr)
library(ggplot2)
library(igraph)
library(reshape2)
library(scales)
library(tidyverse)
library(GENIE3)

theme_set(theme_light())
```

# Helper functions

```{r}
grn.heatmap <- function(dist_mat){
  library(reshape2)
  
  p <- ggplot(data = melt(dist_mat), aes(x = Var1, y = Var2, fill = value)) +
    geom_tile(color = "gray") + coord_equal() +
    scale_fill_gradient2(low = "white", high = 'steelblue') +
    theme(axis.text.x = element_text(angle = 90, hjust = 1),
          panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.border = element_blank(), axis.ticks = element_blank())
  
  return(p)
}
```

This function isn't working right now. Will revisit later.

```{r}
grn.matrix <- function(experiment_data, gene_col, sample_cols){
  library(GENIE3)

  expression_data <- experiment_data %>%
    column_to_rownames(var = gene_col) %>%
    select(sample_cols - 1) %>%
    as.matrix() 
  
  distmat <- GENIE3(expression_data)

  return(distmat)
}
```


# Metabolomics data

```{r}
metabolomics <- read.delim("./data/Metabolomics2.txt")
```

### Building GRN

Can't build it right now since there are duplicate row names.


# Proteomics

```{r}
proteomics <- read.delim("./data/Proteomics2.txt")
```

### Building GRN

```{r}
expr <- proteomics %>% 
  column_to_rownames(var = "gene") %>% 
  select(7:12) %>% 
  as.matrix()

proteomics.mat <- GENIE3(expr)
```


#  Transcriptomics

```{r}
transcriptomics <- read.delim("./data/Transcriptomics2.txt")
```

### Building GRN

Takes a long time to run. Will run on HPC.

```{r}
# expr <- transcriptomics %>% 
  column_to_rownames(var = "Unique_ID") %>% 
  select(8:13) %>% 
  as.matrix()

# transcriptomics.mat <- GENIE3(expr)
```

### Reading in network from GENIE3 adjaceny matrices.

```{r}
transcriptomics.network <- graph_from_adjacency_matrix(transcriptomics.mat)
```








